src/data-access/repositories/user-repository.ts

```diff
@@ -0,0 +1,197 @@
+import { ApolloError } from 'apollo-server-lambda';
+import { DocumentClient, QueryOutput, ScanOutput } from 'aws-sdk/clients/dynamodb';
+import { Maybe } from 'graphql/jsutils/Maybe';
+import { UsersDBConfiguration } from '../../configuration/users-db';
+
+export interface UserRepository {
+	createUser(data: UserInput): Promise<UserModel>;
+	updateUser(id: string, data: UserInput): Promise<UserModel>;
+	deleteUser(id: string): Promise<UserModel>;
+	getUser(id: string): Promise<UserModel>;
+	listUsers(query: Maybe<string>, limit: number, cursor: Maybe<string>): Promise<UserPageModel>;
+}
+
+interface ValidatedInput extends UserInput {
+	formattedDateOfBirth: string | null;
+}
+
+export class DynamoDBUserRepository implements UserRepository {
+	constructor(private database: AWS.DynamoDB.DocumentClient, private config: UsersDBConfiguration) {}
+
+	async createUser(data: UserInput): Promise<UserModel> {
+		if (!data.name) {
+			throw new UserInputError('The name must be specified');
+		}
+
+		if (!data.dob) {
+			throw new UserInputError('The date of birth must be specified');
+		}
+
+		if (!data.address) {
+			throw new UserInputError('The address must be specified');
+		}
+
+		const validatedInput = this.validateInput(data);
+		const user = {
+			id: uuid(),
+			name: data.name,
+			address: data.address,
+			description: data.description,
+			dob: validatedInput.formattedDateOfBirth,
+			createdAt: moment().toISOString(),
+		} as UserModel;
+
+		await this.database
+			.put({
+				TableName: this.config.tableName,
+				Item: user,
+			})
+			.promise();
+
+		return user;
+	}
+
+	async updateUser(id: string, data: UserInput): Promise<UserModel> {
+		const validatedInput = this.validateInput(data);
+		const user = await this.getUser(id);
+
+		user.name = data.name || user.name;
+		user.address = data.address || user.address;
+		user.dob = validatedInput.formattedDateOfBirth || user.dob;
+		user.description = data.description || user.description;
+		user.updatedAt = moment().toISOString();
+
+		// TODO: Change this code to use a update operation
+		await this.database
+			.put({
+				TableName: this.config.tableName,
+				Item: user,
+			})
+			.promise();
+
+		return user;
+	}
+
+	async getUser(id: string): Promise<UserModel> {
+		const response = await this.database
+			.get({
+				TableName: this.config.tableName,
+				Key: { id },
+			})
+			.promise();
+
+		if (!response.Item) {
+			throw new ApolloError(`Could not find user with id ${id}`);
+		}
+
+		return response.Item as UserModel;
+	}
+
+	async deleteUser(id: string): Promise<UserModel> {
+		const result = await this.database
+			.delete({
+				TableName: this.config.tableName,
+				Key: { id },
+			})
+			.promise();
+
+		return result.Attributes as UserModel;
+	}
+
+	async listUsers(query: Maybe<string>, limit: number, cursor: Maybe<string>): Promise<UserPageModel> {
+		// TODO: This API would be more powerful if it used a full text search engine. In production we should
```

Comment:
How do you manage TODO in a real-world project?

Reply:
I usually create an issue on a project management software (in my current work, we use JIRA to do that) when it is something that should be "leaked" to another issue. If it is an issue I am currently developing and I just want to keep track of what I need to do before I finish it, I use Notion (I like to use it to manage my activities in general). In the case of this project, I decided to leave these TODO comments to highlight that I am aware that there might be better ways to do what I was trying to do.